geode-for-redis/src/main/java/org/apache/geode/redis/internal/eventing/EventDistributor.java

```diff
@@ -0,0 +1,128 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Timer;
+import java.util.TimerTask;
+
+import org.apache.logging.log4j.Logger;
+
+import org.apache.geode.annotations.VisibleForTesting;
+import org.apache.geode.cache.partition.PartitionListenerAdapter;
+import org.apache.geode.logging.internal.log4j.api.LogService;
+import org.apache.geode.redis.internal.commands.RedisCommandType;
+import org.apache.geode.redis.internal.data.RedisKey;
+
+public class EventDistributor extends PartitionListenerAdapter {
+
+  private static final Logger logger = LogService.getLogger();
+  private final Map<RedisKey, List<EventListener>> listeners = new HashMap<>();
+  private final Map<EventListener, TimerTask> timerTasks = new HashMap<>();
+  private final Timer timer = new Timer("GeodeForRedisEventTimer", true);
+  private int keysRegistered = 0;
+
+  public synchronized void registerListener(EventListener listener) {
+    for (RedisKey key : listener.keys()) {
+      listeners.computeIfAbsent(key, k -> new ArrayList<>()).add(listener);
+    }
+
+    if (listener.getTimeout() != 0) {
+      scheduleTimeout(listener, listener.getTimeout());
+    }
+
+    keysRegistered += listener.keys().size();
+  }
+
+  public synchronized void fireEvent(RedisCommandType command, RedisKey key) {
+    if (listeners.isEmpty()) {
+      return;
+    }
+
+    List<EventListener> listenerList = listeners.get(key);
+    if (listenerList == null) {
+      return;
+    }
+
+    for (EventListener listener : listenerList) {
+      if (listener.process(command, key) == EventResponse.REMOVE_AND_STOP) {
+        removeListener(listener);
+        break;
+      }
+    }
+  }
+
+  /**
+   * The total number of keys registered by all listeners (includes duplicates).
+   */
+  @VisibleForTesting
+  public int size() {
+    return keysRegistered;
+  }
+
+  @Override
+  public synchronized void afterBucketRemoved(int bucketId, Iterable<?> keys) {
+    Set<EventListener> resubmittingList = new HashSet<>();
+    for (Map.Entry<RedisKey, List<EventListener>> entry : listeners.entrySet()) {
+      if (entry.getKey().getBucketId() == bucketId) {
+        resubmittingList.addAll(entry.getValue());
+      }
+    }
+
+    resubmittingList.forEach(x -> {
+      x.resubmitCommand();
+      removeListener(x);
+    });
+  }
+
+  private synchronized void removeListener(EventListener listener) {
+    boolean listenerRemoved = false;
+    for (RedisKey key : listener.keys()) {
+      List<EventListener> listenerList = listeners.get(key);
+      listenerRemoved |= listenerList.remove(listener);
+      if (listenerList.isEmpty()) {
+        listeners.remove(key);
+      }
+    }
+
+    if (listenerRemoved) {
+      keysRegistered -= listener.keys().size();
+      if (listener.getTimeout() > 0) {
+        timerTasks.remove(listener).cancel();
+      }
+    }
+  }
+
+  private void scheduleTimeout(EventListener listener, long timeout) {
+    TimerTask task = new TimerTask() {
```

Comment:
Another nice thing about using ScheduledThreadPoolExecutor is you can just have the EventListener implement Runnable. You will need to store the Future returned from calling submit somewhere so that you can call cancel on that Future if the timeout is hit.